File: src/24_ReverseList/24_ReverseList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

import os
import sys

sys.path.append(os.path.split(sys.path[0])[0])

from Utilities import list_utils
from Utilities.list_utils import ListNode


class Solution:
    def reverseList(self, head: ListNode, use_iter=True) -> ListNode:
        if use_iter:
            return self.reverseList_Iteratively(head)
        else:
            return self.reverseList_Recursively(head, None)

    # 基于迭代
    def reverseList_Iteratively(self, head: ListNode) -> ListNode:
        cur, pre = head, None
        while cur:
            cur.next, pre, cur = pre, cur, cur.next
        return pre

    # 基于递归
    def reverseList_Recursively(self, cur: ListNode, pre: ListNode) -> ListNode:
        if cur is None:
            return pre
        newHead = self.reverseList_Recursively(cur.next, cur)
        cur.next = pre
        return newHead


class Test:
    def __init__(self):
        self.solution = Solution()

    def checkResult(self, head: ListNode) -> None:
        print("The original list is:")
        list_utils.printList(head)

        reversed_head = self.solution.reverseList(head)
        print("The reversed list is:")
        list_utils.printList(reversed_head)
        print("\n")

    def Test1(self) -> None:
        vals = [1,2,3,4,5]
        nodes = list_utils.createListNodesVec(vals)
        list_utils.connectListNodesVec(nodes)

        self.checkResult(nodes[0])

    def Test2(self) -> None:
        node1 = list_utils.createListNode(1)
        self.checkResult(node1)

    def Test3(self) -> None:
        self.checkResult(None)

    def runTest(self):
        self.Test1()
        self.Test2()
        self.Test3()


if __name__ == "__main__":
    test = Test()
    test.runTest()

File: src/24_ReverseList/24_ReverseList.cpp
// 面试题24：反转链表
// 题目：定义一个函数，输入一个链表的头结点，反转该链表并输出反转后链表的
// 头结点。

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(nullptr) {}
 * };
 */

#include <iostream>

#include "../Utilities/List.h"

using namespace std;

class Solution {
   public:
    ListNode* reverseList( ListNode* pHead, bool use_iter = true ) {
        if ( use_iter )
            return reverseList_Iteratively( pHead );
        else
            return reverseList_Recursively( pHead, nullptr );
    }

    // 基于迭代
    ListNode* reverseList_Iteratively( ListNode* pHead ) {
        ListNode *cur = pHead, *pre = nullptr;
        while ( cur != nullptr ) {
            // cur->next, pre, cur = pre, cur, cur->next
            ListNode* tmp = cur->next;
            cur->next = pre;
            pre = cur;
            cur = tmp;
        }
        return pre;
    }

    // 基于递归
    ListNode* reverseList_Recursively( ListNode* cur, ListNode* pre ) {
        // 终止条件
        if ( cur == nullptr ) return pre;
        // 递归后继节点
        ListNode* newHead = reverseList_Recursively( cur->next, cur );
        cur->next = pre;  // 修改节点引用指向
        return newHead;   // 返回倒转后的头节点
    }
};

class Test {
   public:
    void checkResult( ListNode* pHead ) {
        cout << "\nThe original list is: \n";
        printList( pHead );

        ListNode* pReversedHead = solution.reverseList( pHead );
        cout << "The reversed list is:\n";
        printList( pReversedHead );
        cout << "\n";

        destroyList( pReversedHead );
    }

    // 输入的链表有多个结点
    void Test1() {
        ValVec vals{ 1, 2, 3, 4, 5 };
        ListNodeVec nodes = createListNodesVec( vals );
        connectListNodesVec( nodes );
        checkResult( nodes[ 0 ] );
    }

    // 输入的链表只有一个结点
    void Test2() {
        ListNode* pNode1 = createListNode( 1 );
        checkResult( pNode1 );
    }

    // 输入空链表
    void Test3() { checkResult( nullptr ); }

    void runTest() {
        Test1();
        Test2();
        Test3();
    }

   private:
    Solution solution;
};

int main() {
    Test test;
    test.runTest();
    return 0;
}
